src/transactions/BumpExpirationOpFrame.cpp

```diff
@@ -0,0 +1,167 @@
+{
+    return header.ledgerVersion >= 20;
+}
+
+bool
+BumpExpirationOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    throw std::runtime_error("BumpExpirationOpFrame::doApply needs Config");
+}
+
+bool
+BumpExpirationOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
+                               Hash const& sorobanBasePrngSeed)
+{
+    BumpExpirationMetrics metrics(app.getMetrics());
+
+    auto const& resources = mParentTx.sorobanResources();
+    auto const& footprint = resources.footprint;
+
+    UnorderedMap<LedgerKey, uint32_t> originalExpirations;
+
+    for (auto const& lk : footprint.readOnly)
+    {
+        // When we move to use EXPIRATION_EXTENSIONS, this should become a
+        // loadWithoutRecord, and the metrics should be updated.
+        auto ltxe = ltx.load(lk);
+        if (ltxe)
+        {
+            auto keySize = xdr::xdr_size(lk);
+            auto entrySize = xdr::xdr_size(ltxe.current());
+            metrics.mLedgerWriteByte += keySize;
+            metrics.mLedgerWriteByte += entrySize;
+
+            // Divide the limit by 2 for the metadata check since the previous
+            // state is also included in the meta.
+            if (resources.extendedMetaDataSizeBytes / 2 <
+                    metrics.mLedgerWriteByte ||
+                resources.readBytes < metrics.mLedgerWriteByte ||
+                resources.writeBytes < metrics.mLedgerWriteByte)
+            {
+                innerResult().code(BUMP_EXPIRATION_RESOURCE_LIMIT_EXCEEDED);
+                return false;
+            }
+
+            originalExpirations.emplace(LedgerEntryKey(ltxe.current()),
+                                        getExpirationLedger(ltxe.current()));
+
+            auto ledgerSeq = ltx.loadHeader().current().ledgerSeq;
+            uint32_t bumpTo = UINT32_MAX;
+            if (UINT32_MAX - ledgerSeq > mBumpExpirationOp.ledgersToExpire())
+            {
+                bumpTo = ledgerSeq + mBumpExpirationOp.ledgersToExpire();
+            }
+
+            if (getExpirationLedger(ltxe.current()) < bumpTo)
+            {
+                setExpirationLedger(ltxe.current(), bumpTo);
+            }
+        }
+    }
+
+    mParentTx.pushInitialExpirations(std::move(originalExpirations));
+
+    innerResult().code(BUMP_EXPIRATION_SUCCESS);
+    return true;
+}
+
+bool
+BumpExpirationOpFrame::doCheckValid(SorobanNetworkConfig const& config,
+                                    uint32_t ledgerVersion)
+{
+    auto const& footprint = mParentTx.sorobanResources().footprint;
+    if (!footprint.readWrite.empty())
+    {
+        innerResult().code(BUMP_EXPIRATION_MALFORMED);
+        return false;
+    }
+
+    for (auto const& lk : footprint.readOnly)
+    {
+        if (!isSorobanEntry(lk))
+        {
+            innerResult().code(BUMP_EXPIRATION_MALFORMED);
+            return false;
+        }
+    }
+
+    // check for duplicates
+    UnorderedSet<LedgerKey> set;
```

Comment:
Interestingly, we seemingly allow duplicates for the invoke host function op. Maybe this check should move to tx level and be performed for all the Soroban txs?

Reply:
Yeah we can put it in `commonValidPreSeqNum` with the other resource validations.

Reply:
Updated to forbid duplicates across readOnly and readWrite sets.